File: ios/Classes/FullScreenAdCreator.swift
/*
 * This file is a part of the Yandex Advertising Network
 *
 * Version for Flutter (C) 2023 YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */

import Flutter
import YandexMobileAds

@MainActor
final class FullScreenAdCreator {

    let messenger: FlutterBinaryMessenger

    private static var nextIdByChannelName: [String: Int] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
    }

    func createAppOpenAd(ad: AppOpenAd) -> Int {
        let viewController = AppOpenAdViewController()
        ad.delegate = viewController
        return createAd(channelName: appOpenAdChannelName, viewController: viewController) {
            AppOpenAdCommandProvider(ad: ad, onDestroy: $0, adViewController: viewController)
        }
    }

    func createInterstitialAd(ad: InterstitialAd) -> Int {
        let viewController = InterstitialAdViewController()
        ad.delegate = viewController
        return createAd(channelName: interstitialAdChannelName, viewController: viewController) {
            InterstitialAdCommandProvider(ad: ad, onDestroy: $0, adViewController: viewController)
        }
    }

    func createRewardedAd(ad: RewardedAd) -> Int {
        let viewController = RewardedAdViewController()
        ad.delegate = viewController
        return createAd(channelName: rewardedAdChannelName, viewController: viewController) {
            RewardedAdCommandProvider(ad: ad, onDestroy: $0, adViewController: viewController)
        }
    }

    private func createAd(
        channelName: String,
        viewController: BaseFullscreenAdViewController & NSObjectProtocol,
        providerFactory: (_ onDestroy: @escaping () -> Void) -> CommandProvider
    ) -> Int {
        let id = Self.nextIdByChannelName[channelName, default: 0]
        Self.nextIdByChannelName[channelName] = id + 1
        let name = "\(YandexMobileAdsPlugin.channelName).\(channelName).\(id)"
        let methodChannel = FlutterMethodChannel(name: name, binaryMessenger: messenger)
        let eventChannel = FlutterEventChannel(name: "\(name).events", binaryMessenger: messenger)
        let provider = providerFactory {
            methodChannel.setMethodCallHandler(nil)
            eventChannel.setStreamHandler(nil)
        }
        methodChannel.setMethodCallHandler(provider.callHandler)
        eventChannel.setStreamHandler(viewController)
        return id
    }

    private let appOpenAdChannelName = "appOpenAd"
    private let interstitialAdChannelName = "interstitialAd"
    private let rewardedAdChannelName = "rewardedAd"
}

File: android/src/main/kotlin/com/yandex/mobile/ads/flutter/common/FullScreenAdCreator.kt
/*
 * This file is a part of the Yandex Advertising Network
 *
 * Version for Flutter (C) 2023  YANDEX
 *
 * You may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://legal.yandex.com/partner_ch/
 */

package com.yandex.mobile.ads.flutter.common

import com.yandex.mobile.ads.flutter.FullScreenEventListener
import com.yandex.mobile.ads.flutter.YandexMobileAdsPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel

internal class FullScreenAdCreator(
    private val messenger: BinaryMessenger
) {

    fun createFullScreenAd(
        channelName: String,
        listener: FullScreenEventListener,
        onDestroyHandlerProvider: (onDestroy: () -> Unit) -> CommandHandlerProvider,
    ): Int {
        val id = idCount++
        val name = "${YandexMobileAdsPlugin.ROOT}.$channelName.$id"
        val methodChannel = MethodChannel(messenger, name)
        val eventChannel = EventChannel(messenger, "$name.events")
        val provider = onDestroyHandlerProvider {
            methodChannel.setMethodCallHandler(EmptyMethodCallHandler())
            eventChannel.setStreamHandler(null)
        }
        methodChannel.setMethodCallHandler { call, result ->
            provider.commandHandlers[call.method]?.handleCommand(call.method, call.arguments, result)
                ?: result.notImplemented()
        }
        eventChannel.setStreamHandler(listener)
        return id
    }

    companion object {

        private var idCount = 0
    }
}
